gitautosync/pull_from_remote.py

```diff
@@ -0,0 +1,129 @@
+    logger.info('    Branch: {}'.format(branch_name))
+    logger.info('    Repo Dir: {}'.format(repo_dir))
+
+    if not os.path.exists(repo_dir):
+        _initialize_repo(git_url, repo_dir)
+    else:
+        _update_repo(git_url, repo_dir, branch_name)
+
+    logger.info('Pulled from repo: {}'.format(git_url))
+
+
+def _initialize_repo(git_url, repo_dir):
+    """
+    Clones repository.
+    """
+    assert git_url
+
+    logger.info('Repo {} doesn\'t exist. Cloning...'.format(repo_dir))
+    subprocess.check_call(['git', 'clone', git_url, repo_dir])
+    logger.info('Repo {} initialized'.format(repo_dir))
+
+
+def _update_repo(git_url, repo_dir, branch_name):
+    """
+    Update repo by merging local and upstream changes
+    """
+    assert git_url and branch_name
+
+    _reset_deleted_files(repo_dir, branch_name)
+    if _repo_is_dirty(repo_dir):
+        _make_commit(repo_dir, branch_name)
+    _pull_and_resolve_conflicts(git_url, repo_dir, branch_name)
+
+
+def _reset_deleted_files(repo_dir, branch_name):
+    """
+    Runs the equivalent of git checkout -- <file> for each file that was
+    deleted. This allows us to delete a file, hit an interact link, then get a
+    clean version of the file again.
+    """
+    assert branch_name
+
+    cwd = _get_sub_cwd(repo_dir)
+    status = subprocess.check_output(['git', 'status'], cwd=cwd)
+    deleted_files = DELETED_FILE_REGEX.findall(status.decode('utf-8'))
+
+    for filename in deleted_files:
+        subprocess.check_call(['git', 'checkout', '--', _clean_path(filename)], cwd=cwd)
+        logger.info('Resetted {}'.format(filename))
+
+
+def _make_commit(repo_dir, branch_name):
+    """
+    Commit local changes
+    """
+    assert branch_name
+
+    cwd = _get_sub_cwd(repo_dir)
+    subprocess.check_call(['git', 'checkout', branch_name], cwd=cwd)
+    subprocess.check_call(['git', 'add', '-A'], cwd=cwd)
+    subprocess.check_call(['git', 'config', 'user.email', '"gitautopull@email.com"'], cwd=cwd)
+    subprocess.check_call(['git', 'config', 'user.name', '"GitAutoPull"'], cwd=cwd)
+    subprocess.check_call(['git', 'commit', '-m', 'WIP'], cwd=cwd)
+    logger.info('Made WIP commit')
+
+
+def _pull_and_resolve_conflicts(git_url, repo_dir, branch_name):
+    """
+    Git pulls, resolving conflicts with -Xours
+    """
+    assert git_url and branch_name
+
+    logger.info('Starting pull from {}'.format(git_url))
+
+    cwd = _get_sub_cwd(repo_dir)
+    subprocess.check_call(['git', 'checkout', branch_name], cwd=cwd)
+    subprocess.check_call(['git', 'fetch'], cwd=cwd)
+    subprocess.check_call(['git', 'merge', '-Xours', 'origin/{}'.format(branch_name)], cwd=cwd)
+
+    logger.info('Pulled from {}'.format(git_url))
+
+
+def _repo_is_dirty(repo_dir):
+    """
+    Return empty string if repo not dirty.
+    Return non-empty string if repo dirty.
+    """
+    cwd = _get_sub_cwd(repo_dir)
+    p = subprocess.Popen('git diff-index --name-status HEAD -- | grep -e "^M.*$"',
```

Comment:
Avoid using shell=True everywhere if possible - it can cause big security issues (see https://docs.python.org/2/library/subprocess.html#frequently-used-arguments). In this case, if you really needed this, you should just use check_output and parse it in python.